### PR TITLE
Fixed posterior for NormalWishart.

### PR DESCRIPTION
**Pull request by krylea**
The posterior calculations for most of this package are based on the paper Conjugate Bayesian analysis of the Gaussian distribution by Kevin Murphy. This paper has several typos related to the Normal-Wishart distribution. The ConjugatePriors package fixes the typos related to the pdf of the Normal-Wishart (though I would note that right now the pdf function will not run without errors for this distribution, as it relies on a function called IsApproxSymmetric that does not appear to exist - but that is a separate issue), but it does not fix the typos in the posterior. The paper cites the posterior update for the scale matrix as being:
 T_n = T_0 + S + k_0/k_n * n * (mu_0 - xhat)(mu0-xhat)'
The correct update is 
inv(T_n) = inv(T_0) + S + k_0/k_n * n * (mu_0 - xhat)(mu0-xhat)'
See https://stats.stackexchange.com/questions/153241/derivation-of-normal-wishart-posterior/324925#324925 for details.

### CODE REVIEW

**Comment by krylea** [comment]
Just a comment that I have tested this with a MCMC sampler I am using. When I use the existing posterior update, the sampler does not converge, and gives nonsensical results. When I use the update I added here, it converges quickly to the expected answer.

**Comment by mschauer** [comment]
I checked with https://en.wikipedia.org/wiki/Conjugate_prior#Continuous_distributions, so I agree that this is correct. Is there a way that we could interchange last Cholesky und inversion and return a Wishart which is parametrized differently?

**Comment by krylea** [comment]
After looking at this further, it seems like the T that is intended to be stored in the Normal-Wishart object here is the inverse scale matrix for the Wishart, rather than the scale matrix itself (at least judging by the implementation of the pdf and posterior functions). In this case, the existing posterior update and pdf calculation for the scale matrix should be retained, and they are correct as is. However, T should be inverted before passing it to the Wishart in the rand function. (Also, this should probably be better documented, as it is not the standard parametrization of the Normal-Wishart)

**Comment by mschauer** [comment]
Great! This is the original implementation: https://github.com/JuliaStats/Distributions.jl/pull/140/files Maybe @nfoti remembers

**Comment by krylea** [comment]
I believe I updated the PR to reflect the inverse scale matrix parametrization. It would be good to have someone look it over though to make sure I didn't miss anything. I believe what I have now is the same as the original formulation before my pull request, except for a change to the rand function.

**Comment by krylea** [comment]
After checking again, the pdf function actually did need to be changed for this parametrization. I also fixed a syntax error in logpdf, and commented out a line which called the "isapproxsymmetric" function, which is not a function that appears to exist as far as I can tell.

**Comment by krylea** [comment]
Apologies for making so many commits. I have now gone through and thoroughly tested this and am confident it is now right.

**Comment by mschauer** [comment]
Don't worry. Can we cook up a test which shows that this is right and the version on master is wrong?

**Comment by krylea** [comment]
I am working on writing something up, I will hopefully post it later today.

**Comment by mschauer** [comment]
One suggestion would be to use in test that the one-dimensional case falls back to the InverseGamma/Gamma distribution, so they should give corresponding results.

File: src/normalwishart.jl [hunk 1]
```diff
@@ -62,27 +60,28 @@ function logpdf(nw::NormalWishart, x::Vector{T}, Lam::Matrix{T}) where T<:Real
...
         return logp
-
     end
 end
 
 function rand(nw::NormalWishart)
-    Lam = rand(Wishart(nw.nu, nw.Tchol))
-    Lsym = PDMat(Symmetric(inv(Lam) ./ nw.kappa))
-    mu = rand(MvNormal(nw.mu, Lsym))
+    V = PDMat(Symmetric(inv(nw.Tchol)))
```
[the remark this answers is not in the document]
Comment (krylea) [reply src/normalwishart.jl]: Yes. Julia's matrix inversion can lead to some small inaccuracies (of ordr 1e-16 or 1e-18 or so) and thus inverse matrices that SHOULD be symmetric will be slightly asymmetric. This shouldn't be a problem, but the distributions we are passing these two require symmetric matrices, and enforce a check for symmetricity. Julia's base is_hermitian function does not have a relative tolerance, and thus without some way of enforcing that the inverted matrix is symmetric, this will crash.

File: src/normalwishart.jl [hunk 2]
```diff
@@ -62,27 +60,28 @@ function logpdf(nw::NormalWishart, x::Vector{T}, Lam::Matrix{T}) where T<:Real
         hp = 0.5 * p
 
         # Normalization
-        logp = hp*(log(kappa) - Float64(log2π))
+        logp = hp*(log(kappa) - Float64(log(2*π)))
```
[the remark this answers is not in the document]
Comment (krylea) [reply src/normalwishart.jl]: My bad, I think it was giving me errors when I tried to run it. Maybe I just wasn't importing the required package or something.

**Comment by krylea** [comment]
To be honest I am still not entirely sure how best to approach the problem of how to parametrize this - but no matter how we decide it is best to do it, the current implementation needs to be fixed. I apologize for not getting back to you on the testing issue, I have been very busy recently so haven't had time to look at it.

File: src/normalwishart.jl [hunk 3]
```diff
@@ -62,27 +60,28 @@ function logpdf(nw::NormalWishart, x::Vector{T}, Lam::Matrix{T}) where T<:Real
...
         logp -= hnu * logdet(Tchol)
         logp -= hnu * p * log(2.)
-        logp -= lpgamma(p, hnu)
+        logp -= logmvgamma(p, hnu)
```
[the remark this answers is not in the document]
Comment (krylea) [reply src/normalwishart.jl]: There were other areas too where the current logpdf function crashes. I think somewhere in insupport there is a call to a function that doesn't exist. I didn't fix that in this pull request because it was a separate issue, and I forgot to make a new issue for it.

File: src/normalwishart.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (mschauer) [reply src/normalwishart.jl]: We discussed this somewhere, and this was the conclusion, but I cannot find it anymore.

[the remark this answers is not in the document]
Comment (kleinschmidt) [reply src/normalwishart.jl]: I suppose this might have something to do with the missing `isApproxSymmetric` function...

**Comment by kleinschmidt** [comment]
I agree that the parametrization can be fixed later, but tests are critical before we merge.  Even something as simple as calling the functions on a few values would have caught many of these bugs... Thanks for your work on this!

File: src/normalwishart.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (kleinschmidt) [reply src/normalwishart.jl]: Yeah I think it's necessary to add StatsFuns (which also provides `logmvgamma` which is needed elsewhere).

File: src/normalwishart.jl [hunk 4]
```diff
@@ -52,8 +52,6 @@ function logpdf(nw::NormalWishart, x::Vector{T}, Lam::Matrix{T}) where T<:Real
     if !insupport(NormalWishart, x, Lam)
         return -Inf
     else
-        p = length(x)
```
[the remark this answers is not in the document]
Comment (kleinschmidt) [reply src/normalwishart.jl]: alternatively, you could use `nw.dim`

**Comment by kleinschmidt** [comment]
Also, i'm looking the Normal-Inverse-Wishart and seeing a lot of the same problems.  Would you be willing to fix those as well? If not let's merge this after adding some minimal tests (at a minimum just to make sure this code can be executed without errors)

**Comment by kleinschmidt** [comment]
ah I found the source of that function (as well as another one that's used by NIW): https://github.com/JuliaStats/Distributions.jl/blob/2fc223f72a2df4934c47e5701b9afc8119e0448f/src/utils.jl#L117-L137

File: src/normalwishart.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (krylea) [reply src/normalwishart.jl]: logmvgamma is from Distributions. maybe the old lpgamma is a version from StatsFuns?

File: src/normalwishart.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (krylea) [reply src/normalwishart.jl]: I think this was an accident

File: src/normalwishart.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (krylea) [reply src/normalwishart.jl]: The problem is that the symmetry checks are not in the ConjugatePriors package, they are in various places in Distributions. So even if we make our symmetry checks sensitive to a relative tolerance, the checks there won't be. To be honest I have no idea why Julia won't just add a relative tolerance to its is_hermitian function given that the symmetry errors are coming from Julia's own matrix inversion routines, but I was told that that isn't going to happen...

**Comment by krylea** [comment]
I'll try to get to this today and look over everything again. Sorry I've been dragging my heels, I've just been caught up in other work.

**Comment by kleinschmidt** [comment]
No worries! I'm in the same boat :)

File: src/normalwishart.jl [hunk 5]
```diff
@@ -40,7 +40,7 @@ end
 function insupport(::Type{NormalWishart}, x::Vector{T}, Lam::Matrix{T}) where T<:Real
     return (all(isfinite(x)) &&
            size(Lam, 1) == size(Lam, 2) &&
-           isApproxSymmmetric(Lam) &&
+          #isApproxSymmmetric(Lam) &&
```
Comment (kleinschmidt) [inline src/normalwishart.jl]: This (and `hasCholesky`) are in Distributions, so can be fixed by qualifying with `Distributions.`

Comment (kleinschmidt) [reply src/normalwishart.jl]: maybe worth moving them over here since I don't think they're used in Distributions anymore...

File: src/normalwishart.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (kleinschmidt) [reply src/normalwishart.jl]: I think Distributions gets it from StatsFuns: https://github.com/JuliaStats/StatsFuns.jl/blob/e66dd973650c375bc1739c820e5b96bb5bd000a8/src/misc.jl#L3-L15

**Comment by kleinschmidt** [comment]
@krylea I have a bit of time to work on this now actually; is it okay if I push more commits to this PR?

**Comment by krylea** [comment]
Go for it!

**Comment by krylea** [comment]
After taking another look, I believe the alternate parameterization we were considering here is actually the parameterization for the Normal INVERSE Wishart. I returned to the original parameterization, and added the inversions that were missing in the base code.

**Comment by mschauer** [comment]
What is the status here?

**Comment by matbesancon** [comment]
ping @krylea

**Comment by mschauer** [comment]
@johnczito Would this fall into your area of expertise, be interesting for you?

**Comment by johnczito** [comment]
Sure. I guess the first thing that comes to mind is that `NormalWishart` should have a field `W` that simply stores `Wishart(nu, Tchol)`. Also, why shouldn't `logpdf` just be this?

```julia
function logpdf(nw::NormalWishart, x::Vector{T}, Lam::Matrix{T}) where T<:Real
    Lsym = PDMat(Symmetric(inv(Lam) ./ nw.kappa))
    logpdf(Wishart(nw.nu, nw.Tchol), Lam) + logpdf(MvNormal(nw.mu, Lsym), x)
end
```

**Comment by johnczito** [comment]
^ That at least suggests a unit test you'd want to implement, in addition to comparing against the `NormalGamma`.
